napi-inl.h

```diff
@@ -4356,6 +4394,363 @@ inline void AsyncWorker::OnWorkComplete(Napi::Env /*env*/, napi_status status) {
+  ThreadSafeFunctionEx<ContextType, DataType, CallJs> tsfn;
+
+  auto *finalizeData = new details::ThreadSafeFinalize<ContextType, Finalizer,
+                                                       FinalizerDataType>(
+      {data, finalizeCallback});
+  napi_status status = napi_create_threadsafe_function(
+      env, callback, nullptr, String::From(env, resourceName), maxQueueSize,
+      initialThreadCount, finalizeData,
+      details::ThreadSafeFinalize<ContextType, Finalizer, FinalizerDataType>::
+          FinalizeFinalizeWrapperWithDataAndContext,
+      context, CallJsInternal, &tsfn._tsfn);
+  if (status != napi_ok) {
+    delete finalizeData;
+    NAPI_THROW_IF_FAILED(env, status,
+                         ThreadSafeFunctionEx<ContextType, DataType, CallJs>());
+  }
+
+  return tsfn;
+}
+
+// static, with: Callback [passed] Resource [passed] Finalizer [passed]
+template <typename ContextType, typename DataType,
+          void (*CallJs)(Napi::Env, Napi::Function, ContextType *, DataType *)>
+template <typename CallbackType, typename ResourceString, typename Finalizer,
+          typename FinalizerDataType>
+inline ThreadSafeFunctionEx<ContextType, DataType, CallJs>
+ThreadSafeFunctionEx<ContextType, DataType, CallJs>::New(
+    napi_env env, CallbackType callback, const Object &resource,
+    ResourceString resourceName, size_t maxQueueSize, size_t initialThreadCount,
+    ContextType *context, Finalizer finalizeCallback, FinalizerDataType *data) {
+  ThreadSafeFunctionEx<ContextType, DataType, CallJs> tsfn;
+
+  auto *finalizeData = new details::ThreadSafeFinalize<ContextType, Finalizer,
+                                                       FinalizerDataType>(
+      {data, finalizeCallback});
+  napi_status status = napi_create_threadsafe_function(
+      env, details::DefaultCallbackWrapper<CallbackType, ThreadSafeFunctionEx<ContextType, DataType, CallJs>>(env, callback), resource,
+      String::From(env, resourceName), maxQueueSize, initialThreadCount,
+      finalizeData,
+      details::ThreadSafeFinalize<ContextType, Finalizer, FinalizerDataType>::
+          FinalizeFinalizeWrapperWithDataAndContext,
+      context, CallJsInternal, &tsfn._tsfn);
+  if (status != napi_ok) {
+    delete finalizeData;
+    NAPI_THROW_IF_FAILED(env, status,
+                         ThreadSafeFunctionEx<ContextType, DataType, CallJs>());
+  }
+
+  return tsfn;
+}
+
+template <typename ContextType, typename DataType,
+          void (*CallJs)(Napi::Env, Napi::Function, ContextType *, DataType *)>
+inline ThreadSafeFunctionEx<ContextType, DataType,
+                            CallJs>::ThreadSafeFunctionEx()
+    : _tsfn() {}
+
+template <typename ContextType, typename DataType,
+          void (*CallJs)(Napi::Env, Napi::Function, ContextType *, DataType *)>
+inline ThreadSafeFunctionEx<ContextType, DataType, CallJs>::
+    ThreadSafeFunctionEx(napi_threadsafe_function tsfn)
+    : _tsfn(tsfn) {}
```

Comment:
IINM this constructor needs to also receive a `Napi::Env` so `.Ref()` and `.Unref()` can later be called on the resulting `Napi::ThreadSafeFunctionEx`.

Reply:
A test would be nice to make sure that `Un/Ref()`ing a pre-existing `napi_threadsafe_function` works.

Reply:
TSFN[Ex] has its own implementation of `.Ref()` and `.Unref()` that accept the env as a parameter. It's not stored with the TSFN object.

https://github.com/KevinEady/node-addon-api/blob/3405b2bf39793a573d420584ea81a411e5127634/napi.h#L2382

Does this suffice...?

Reply:
@KevinEady I missed that. 👍 on not needing the `env` in the constructor.